### PR TITLE
Implement lazy loading of raw data for local files

### PR DESCRIPTION
**Pull request by marcoross**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## [Unreleased]
 
+### Added
+- Lazy load raw signals from local EDF files. Lazy loading is supported and enabled by default when passing a `str` or `pathlib.Path` object to the `read_edf`function. The behavior can be controlled using the `lazy_load_data` parameters of the `read_edf` function ([#48](https://github.com/the-siesta-group/edfio/pull/48)).
```
Comment (cbrnr) [inline CHANGELOG.md:6]:
```suggestion
- Lazy load raw signals from local EDF files. Lazy loading is supported and enabled by default when passing a `str` or `pathlib.Path` object to the `read_edf` function. The behavior can be controlled using the `lazy_load_data` parameters of the `read_edf` function ([#48](https://github.com/the-siesta-group/edfio/pull/48)).
```

**Comment by cbrnr** [comment]
This looks really good! Not entirely related, but I was wondering if there is some public attribute that reflects the status of the data (i.e. loaded vs. not loaded). Sometimes, it would also be useful to know the total size of the `Edf` object in memory. I don't think this is currently possible, right?

Also, I assume loaded vs. not loaded applies to all signals that are part of an `Edf` objects? That is, either the entire object is loaded or not (as opposed to some selected signals). Correct?

Finally, and this is now least related to this PR (so let me know if you would like me to open a new issue), what is the public way to access the actual data? There's no such thing as `Edf.data`, and `Edf.get_signal()` requires a label. Sorry if I'm missing something obvious!

File: edfio/edf_signal.py [hunk 2]
```diff
@@ -296,6 +314,22 @@ def sampling_frequency(self) -> float:
...
         return self._sampling_frequency
 
+    @property
+    def digital_data(self) -> npt.NDArray[np.int16]:
+        """
+        Numpy array containing the digital (uncalibrated) signal values as 16-bit integers.
+
+        To simplify avoiding inconsistencies between signal data and header fields,
+        individual values in the returned array can not be modified. Use
+        :meth:`EdfSignal.update_data` to update raw signal values with calibrated data.
```
Comment (hofaflo) [inline edfio/edf_signal.py:324]: This is not applicable for `digital_data`, just for `data`, right?

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 ## [Unreleased]
 
+### Added
```
Comment (hofaflo) [inline CHANGELOG.md:5]: Please add an entry mentioning that the digital data can now be accessed publicly

File: edfio/edf.py [hunk 4]
```diff
@@ -871,7 +879,9 @@ def slice_between_seconds(
             else:
                 start_index = start * signal.sampling_frequency
                 stop_index = stop * signal.sampling_frequency
-                signal._digital = signal._digital[int(start_index) : int(stop_index)]
+                signal._digital = signal.digital_data[
```
Comment (hofaflo) [inline edfio/edf.py:882]: I'd like these attribute names to be consistent, my preference would be `[_]digital`, but I'm fine with `[_]digital_data` too

File: edfio/edf.py [hunk 5]
```diff
@@ -1059,10 +1070,19 @@ def read_edf(
...
     Edf
         The resulting :class:`Edf` object.
     """
-    return _read_edf(edf_file)
+    do_lazy_load: bool
+    if lazy_load_data == "auto":
+        do_lazy_load = isinstance(edf_file, (Path, str))
+    else:
+        do_lazy_load = lazy_load_data
+    return _read_edf(edf_file, lazy_load_data=do_lazy_load)
```
Comment (hofaflo) [inline edfio/edf.py:1088]:
```suggestion
    if lazy_load_data == "auto":
        lazy_load_data = isinstance(edf_file, (Path, str))
    return _read_edf(edf_file, lazy_load_data=lazy_load_data)
```

**Comment by hofaflo** [comment]
> This looks really good! Not entirely related, but I was wondering if there is some public attribute that reflects the status of the data (i.e. loaded vs. not loaded).

Yes, this could be useful, I'm not sure how to best show it though, since (with this PR) is done on a per-signal basis :thinking: 

>Sometimes, it would also be useful to know the total size of the `Edf` object in memory. I don't think this is currently possible, right?

Currently not, no... maybe this could be combined with your first suggestion, i.e. display the size of the loaded signals and the total size on disk? E.g. `<Edf 20 signals 0 annotations 50/100 MB loaded>`

> Also, I assume loaded vs. not loaded applies to all signals that are part of an `Edf` objects? That is, either the entire object is loaded or not (as opposed to some selected signals). Correct?

With the implementation suggested in this PR, it could be anything in between as well.

> Finally, and this is now least related to this PR (so let me know if you would like me to open a new issue), what is the public way to access the actual data? There's no such thing as `Edf.data`, and `Edf.get_signal()` requires a label. Sorry if I'm missing something obvious!

Right, currently there's nothing like that. While this could be nice for recordings with uniform sampling frequencies (just return a 2d-array), I'm not sure how to best treat those with differing ones (a list of arrays? fail? ...?). What use case are you thinking about for this? (-> a new issue would be nice here, yes)

**Comment by cbrnr** [comment]
I've created a new issue to discuss accessing the data array in #49.

Regarding the other points, I think it would be nice not to overcomplicate the API. So you are saying it is currently possible to have some signals loaded in memory and some not, because each signal is treated as a separate `EdfSignal` (which is basically a memory-mapped NumPy array)? I'm thinking that for users, it might be useful to know how much memory their entire `Edf` object currently uses.

On the other hand, when loading an EDF file, either no signals or all signals are loaded in memory, so there is no way to influence individual signals being loaded with `read_edf()`. I guess the only way to do load individual signals (from a lazy object) is to access a signal, but then the question is which methods trigger/require loading the data?

**Comment by hofaflo** [comment]
> [...] each signal is treated as a separate `EdfSignal` (which is basically a memory-mapped NumPy array)? 

Exactly, with this PR the array stays memory mapped until the data is accessed for the first time.

> I'm thinking that for users, it might be useful to know how much memory their entire `Edf` object currently uses.

Definitely! Is the suggestion for the extended repr in my above comment more or less what you're thinking about here?

> [...] which methods trigger/require loading the data?

With the currently suggested implementation, this would be
- `EdfSignal.data`
- `EdfSignal.digital_data`
- `EdfSignal.update_data()`
- `Edf.update_data_record_duration()` (all signals)
- `Edf.write()` (all signals)
- `Edf.annotations` (all annotation signals)
- `Edf.starttime` (timekeeping annotation signal)

**Comment by cbrnr** [comment]
> Definitely! Is the suggestion for the extended repr in my above comment more or less what you're thinking about here?

Yes, this would be useful! I'd also expose this in an attribute for convenience. Plus, each underlying `EdfSignal` should also show its memory consumption in its repr (and attribute).

> With the currently suggested implementation, this would be
> ...

What about slicing between seconds or annotations? In any case, if the current memory consumption is available, users will always be able to find out which operation loads the data!

**Comment by hofaflo** [comment]
> Yes, this would be useful! I'd also expose this in an attribute for convenience. Plus, each underlying `EdfSignal` should also show its memory consumption in its repr (and attribute).

:+1: Feel free to open a PR for that, ideally once this one is merged!

> What about slicing between seconds or annotations?

Right, thanks for pointing this out! I'll edit the above list. For non-annotation signals this could even be done without loading the data (as long as the desired slice only contains complete datarecords), though that would complicate the implementation a bit.

**Comment by cbrnr** [comment]
Another question that I had was that if `read_edf()` loads the signal into memory, it actually loads the *digital* data, right? So I'm not sure how surprising it will be that the physical data consumes four times as much?

What I'm trying to say is that this is becoming a bit complicated already, given that the original intention of lazy loading was that people sometimes work with EDF headers only, so they don't need the data. Maybe a simpler option would be to add a separate function `read_edf_header()` instead?

File: edfio/edf_signal.py (hunk 2 above)
Comment (marcoross) [reply edfio/edf_signal.py:324]: This was intended, but I forgot to set the flag to make the array read-only.

Comment (marcoross) [reply edfio/edf_signal.py]: I changed the documentation, and kept the data writeable. Having it read-only caused problems in the current logic regarding EDF+ annotations.

**Review by hofaflo — APPROVED** [review]

**Comment by hofaflo** [comment]
@cbrnr, let's move the discussion about memory consumption information into a new issue!

**Comment by cbrnr** [comment]
Sure! I just thought that maybe the current implementation is not even necessary because it adds too much complexity...

**Comment by hofaflo** [comment]
For the original use case described in #47 you're definitely right :D However, it's a nice way to also speed up working with only a (small) subset of signals.

**Pull request merged**